Sources/Basics/Observability.swift

```diff
@@ -0,0 +1,400 @@
+        var metadata = metadata
+        // FIXME: this brings in the TSC API still
+        if let errorProvidingLocation = error as? DiagnosticLocationProviding, let diagnosticLocation = errorProvidingLocation.diagnosticLocation {
+            metadata = metadata ?? DiagnosticsMetadata()
+            metadata?.stringLocation = diagnosticLocation.description
+        }
+
+        let message: String
+        // FIXME: this brings in the TSC API still
+        // FIXME: string interpolation seems brittle
+        if let diagnosticData = error as? DiagnosticData {
+            message = "\(diagnosticData)"
+        } else if let convertible = error as? DiagnosticDataConvertible {
+            message = "\(convertible.diagnosticData)"
+        } else {
+            message = "\(error)"
+        }
+
+        self.emit(severity: .error, message: message, metadata: metadata)
+    }
+
+    public func emit(warning message: String, metadata: DiagnosticsMetadata? = .none) {
+        self.emit(severity: .warning, message: message, metadata: metadata)
+    }
+
+    public func emit(info message: String, metadata: DiagnosticsMetadata? = .none) {
+        self.emit(severity: .info, message: message, metadata: metadata)
+    }
+
+    public func trap<T>(_ closure: () throws -> T) -> T? {
+        do  {
+            return try closure()
+        } catch {
+            self.emit(error)
+            return nil
+        }
+    }
+}
+
+// MARK: - DiagnosticsMetadata
+
+// designed after https://github.com/apple/swift-distributed-tracing-baggage
+
+/// Provides type-safe access to the DiagnosticsMetadata's values.
+/// This API should ONLY be used inside of accessor implementations.
+///
+/// End users should use "accessors" the key's author MUST define rather than using this subscript, following this pattern:
+///
+///     extension DiagnosticsMetadata {
+///       var testID: String? {
+///         get {
+///           self[TestIDKey.self]
+///         }
+///         set {
+///           self[TestIDKey.self] = newValue
+///         }
+///       }
+///     }
+///
+///     enum TestIDKey: DiagnosticsMetadataKey {
+///         typealias Value = String
+///     }
+///
+/// This is in order to enforce a consistent style across projects and also allow for fine grained control over
+/// who may set and who may get such property. Just access control to the Key type itself lacks such fidelity.
+///
+/// Note that specific baggage and context types MAY (and usually do), offer also a way to set baggage values,
+/// however in the most general case it is not required, as some frameworks may only be able to offer reading.
+
+// FIXME: we currently requires that Value conforms to CustomStringConvertible which sucks
+// ideally Value would conform to Equatable but that has generic requirement
+// luckily, this is about to change so we can clean this up soon
+public struct DiagnosticsMetadata: Equatable {
```

Comment:
Is this intended to store contextual as well as location information?  Locations might of course be scoped, e.g. an error on this like of this manifest file of this package, etc.

Reply:
right, I was trying to get away from needing different abstractions for location and for metadata, and was hoping we can compute or otherwise express the location as piece of metadata at a know key. right not the location metadata key points to a simple string, but we can extend that to be something nested. wdyt?

Reply:
That makes sense, especially since the location is a lot more complicated in some cases that the naïve one in TSC.  I think that location and context are also potentially different things.

In Xcode the diagnostics are not isolated, but rather emitted by adding them to a log recorder, which may in turn be a child of a different log recorder, etc.  This provides an implicit context for where in the task tree the diagnostic occurred (e.g. "while evaluating manifest X for package Y as part of a user action to update the package versions, etc).

That sense of "where in the course of events" the diagnostic fits in makes sense to provide as context, I think.

But then there's still the other notion of relevant location, which I think is what we're talking about here, and of which there could be multiple ("here's the conflicting declaration", "here's the previous declaration", etc).  Those seem perfect as metadata.

So maybe metadata (or "annotations" as I think of them) and context are different things.  Or maybe they are both stored as metadata but with keys of different identity and type.  I do like the typed-key approach, but I think we might want to build some context system on top of it (later).